src/loader/ImGui.cpp

```diff
@@ -0,0 +1,492 @@
+}
+
+void CloseModal()
+{
+    const ImGuiContext& g = *GImGui;
+    const ImGuiWindow* window = g.CurrentWindow;
+
+	CloseModal(window->Name);
+}
+
+void OpenWindow(
+	const std::function<bool()>& draw,
+	const std::string& label
+)
+{
+	s_viewports.emplace(label, draw);
+}
+
+bool AddMainPanel(const std::string& name, const std::function<void()>& callback)
+{
+	const auto& [_, added] = s_panels.emplace(name, callback);
+	return added;
+}
+
+bool RemoveMainPanel(const std::string& name)
+{
+	return s_panels.erase(name) > 0;
+}
+
+bool AddContextGroup(const std::string& name, const std::function<void()>& callback)
+{
+	s_menus.emplace_back(name, callback);
+	return true;
+}
+
+bool RemoveContextGroup(const std::string& name)
+{
+	if (const auto it = std::find_if(
+		s_menus.begin(), s_menus.end(),
+		[&name](const auto& pair) { return pair.first == name; });
+		it != s_menus.end())
+	{
+		s_menus.erase(it);
+		return true;
+	}
+
+	return false;
+}
+
+void MaybeShowContextMenu();
+void Run(const std::function<bool()>& mainLoop)
+{
+	LauncherImGui::Backend::Init(hMainWnd);
+
+	auto draw_main = []()
+	{
+		for (auto it = s_viewports.begin(); it != s_viewports.end();)
+		{
+			if (!it->second())
+				it = s_viewports.erase(it);
+			else
+				++it;
+		}
+
+		for (auto it = s_dimRatios.begin(); it != s_dimRatios.end();)
+		{
+			if (const auto viewport = static_cast<ImGuiViewportP*>(ImGui::FindViewportByID(it->first));
+				viewport == nullptr)
+				it = s_dimRatios.erase(it);
+			else
+			{
+				if (DimParentViewport(viewport, it->second))
+					++it;
+				else
+					it = s_dimRatios.erase(it);
+			}
+		}
+
+		// we do this as a popup to let imgui handle styling as a popup. _However_ imgui fails to handle some
+		// things correctly because we aren't popping up over an imgui window, but are instead implicitly
+		// creating a viewport.
+		if (s_contextOpen)
+		{
+			ImGui::OpenPopup("Context Popup");
+			s_contextOpen = false;
+		}
+
+		MaybeShowContextMenu();
+	};
+
+	while (mainLoop())
+	{
+		LauncherImGui::Backend::DrawFrame(draw_main);
+	}
+
+	LauncherImGui::Backend::Cleanup();
+}
+
+bool HandleWndProc(HWND hWnd, uint32_t msg, uintptr_t wParam, intptr_t lParam)
+{
+	return LauncherImGui::Backend::HandleWndProc(hWnd, msg, wParam, lParam);
+}
+
+void OpenMainWindow()
+{
+	OpenWindow(
+		[]()
+		{
+			ImGui::PushStyleVar(ImGuiStyleVar_ChildRounding, 5.f);
+
+			bool is_open = true;
+
+			const auto monitor = ImGui::GetViewportPlatformMonitor(ImGui::GetMainViewport());
+			ImGui::SetNextWindowSize({ 640.f, 480.f }, ImGuiCond_FirstUseEver);
+			ImGui::SetNextWindowPos(monitor->WorkPos + (monitor->WorkSize * 0.5f), ImGuiCond_FirstUseEver, { 0.5f, 0.5f });
+
+			if (ImGui::Begin("MacroQuest", &is_open))
+			{
+				static std::optional<const std::function<void()>*> current_callback;
+
+				ImGui::BeginChild("Window Selection", ImVec2(ImGui::GetContentRegionAvail().x * 0.2f, 0), ImGuiChildFlags_Border, ImGuiWindowFlags_HorizontalScrollbar);
```

Comment:
Check out the ImGuiTreePanelWindow which generalizese this design. The column width should be a fixed size.